rudder-core/src/main/scala/com/normation/rudder/services/policies/nodeconfig/NodeConfigurationServiceImpl.scala

```diff
@@ -100,54 +98,75 @@ class DetectChangeInNodeConfiguration extends Loggable {
 
           //draftid in one and not the other are new,
           //for the one in both, check both ruleId and cacheValue
 
-          ((currentDrafts.keySet ++ targetDrafts.keySet).map(id => (currentDrafts.get(id), targetDrafts.get(id))).flatMap {
-            case (None, None) => //should not happen
-              Set[RuleId]()
-            case (Some(PolicyCache(ruleId, _, _)), None) =>
+          ((currentDrafts.keySet ++ targetDrafts.keySet).map(id => (id, currentDrafts.get(id), targetDrafts.get(id))).flatMap {
+            case (ruleId, None, None) => //should not happen
+              Set.empty[RuleId]
+
+            /*
+             * Here, the entire rule is no more applied to the node: we
+             * don't update the serial
+             */
+            case (ruleId, Some(drafts), None) =>
               logger.trace(s"`-> rule with ID '${ruleId.value}' was deleted")
-              Set(ruleId)
-            case (None, Some(PolicyCache(ruleId, _, _))) =>
-              logger.trace(s"`-> rule with ID '${ruleId.value}' was added, skipping it")
-              // Ignoring nodes addition to Rules. The node will be updated, thanks to the check on change of hash of node config
-              Set[RuleId]()
-            case (Some(PolicyCache(r0, d0, c0)), Some(PolicyCache(r1, d1, c1))) =>
-              //d0 and d1 are equals by construction, but keep them for future-proofing
-              if(d0 == d1) {
-                if(
-                   //check that the rule is the same
-                      r0 == r1
-                   //and that the policy draft is the same (it's cache value, actually)
-                   && c0 == c1
-
-                ) {
-                  Set[RuleId]() //no modification
-                } else {
-                  logger.trace(s"`-> there was a change in the promise with draft ID '${d0.value}'")
-                  Set(r0,r1)
+              Set.empty[RuleId]
+
+            /*
+             * Here, the entire rule starts to be applied to the node: we
+             * don't update the serial.
+             */
+            case (ruleId, None, Some(drafts)) =>
+              logger.trace(s"`-> rule with ID '${ruleId.value}' was added")
+              Set.empty[RuleId]
+
+            /*
+             * That case is the interesting one: the rule is present in current
+             * and target configuration for the node.
+             * We check if all the drafts are the same (i.e same directives applied,
+             * and no modification in directives). On any change, update the serial.
+             */
+            case (ruleId, Some(cDrafts), Some(tDrafts)) =>
+
+              //here, we can just compare the sets for equality
+
+              if(cDrafts == tDrafts) {
+                Set.empty[RuleId]
+              } else {
+                val diff = ((cDrafts &~ tDrafts) ++ (tDrafts &~ cDrafts)).map { case PolicyCache(RuleId(rid), Cf3PolicyDraftId(did), _) =>
```

Comment:
It's the diff on two set to get what directives changed - ha, and actually I could use --, it could be clearer than &~